File: leetcode/go/_141/main.go
package main

import "fmt"

func main() {
	fmt.Println(hasCycle(nil))
}

type ListNode struct {
	Val  int
	Next *ListNode
}

func hasCycle(head *ListNode) bool {
	mp := make(map[*ListNode]struct{})
	for head != nil && head.Next != nil {
		if _, exist := mp[head.Next]; exist {
			return true
		}
		mp[head.Next] = struct{}{}
		head = head.Next
	}
	return false
}

File: leetcode/python/_141/main.py
from typing import Optional

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow, fast = head, head
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
            if slow == fast: return True
        else: return False

print(Solution().hasCycle(ListNode(3, ListNode(2, ListNode(0, ListNode(-4))))))
